bluetooth: mesh: revisions to support building C++ applications

pabigot opened the pull request:
A couple changes to the headers to avoid violating C++ language requirements.

Using the mesh access macros in C++ also requires #25593.

pabigot committed:
bluetooth: mesh: access: fix initialization order in macros

C++ requires that designated initializers be provided in declaration
order.  Fix the macros that violate this requirement.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Draft until #25593 merged.

jhedberg commented on include/bluetooth/mesh/access.h:
```diff
@@ -257,14 +257,12 @@ struct bt_mesh_model_op {
-	.groups = { [0 ... (CONFIG_BT_MESH_MODEL_GROUP_COUNT - 1)] =         \
-			BT_MESH_ADDR_UNASSIGNED },                           \
-	.user_data = _user_data,                                             \
+	.keys = { BT_MESH_KEY_UNUSED },                                      \
```
This isn't doing the same thing as the original code. `BT_MESH_KEY_UNUSED` is defined as `0xffff`, however what you're doing here is only initialising `.keys[0]` to that value, whereas all other values of the array get initialised to zero.

pabigot replied:
Thanks; I've never used that syntax and didn't realize it was a spread operation.  If so, there may be no way to make this header C++ compatible.

jhedberg replied:
It's a GNU C extension: https://gcc.gnu.org/onlinedocs/gcc/Designated-Inits.html

However, it's a fair question to ask whether we should allow such extensions or require standard C code everywhere. It's super convenient though and probably saves a bit of CPU time since the array doesn't need to be initialised at runtime.

joerchan replied:
I haven't tested it out, but I think the designated initializer can be replaced by some of the macro utils (include/sys/util.h) that we have.

Something like:
```
#define FOO(i, _) BT_MESH_ADDR_UNASSIGNED ,
.groups = { UTIL_LISTIFY(CONFIG_BT_MESH_MODEL_GROUP_COUNT - 1, FOO) }
```
I'm not expert on these though, so you might find a better way in there.

pabigot replied:
@joerchan Thanks, that seems to work.  It does require defining a helper macro to generate the constant; maintainers please provide feedback on how that should be spelled and hidden from the user.

One approach not done here is to move the helper functions to the bottom of the file, outside the doxygen block.  This works, but may be confusing.

joerchan replied:
I think @trond-snekvik is the best person to answer the naming and hiding question. I suppose this should probably be re-usable by users for their own models and so it shouldn't be hidden.

pabigot replied:
OK.  I provided a new one that puts all the internals above the documentation block.  The one with "public" (no `_` affix) could be moved into the block and documented, if it provides something of use outside the helper macros.

pabigot committed:
bluetooth: mesh: access: avoid C99 features unavailable in C++

C99 designated array initializers are not part of the C++ language,
and are not supported in GNU G++.  Replace their use with invocations
of the Zephyr listify macro utility, with ancillary helper functions.

Also fix the initialization order to match declaration order (which
C++ also requires).

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

carlescufi commented:
@pabigot do you want to move this out of draft?

pabigot committed:
bluetooth: mesh: cdb: fix include files and add language directive

Include the files that provide the types and macros used by the
header.  Ensure that all declarations use C language linkage when
included from a file using C++.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
